File: backend/app/catalog_view/gear.py
"""Things bought with nuyen: armor, weapons, general gear, drugs, lifestyles."""

from __future__ import annotations

from ..data_loader import CatalogDict, drug_effect_summary
from ..engine import gear_extra_options


def section(raw: CatalogDict) -> dict:
    return {
        "armor": [
            {
                "id": a["id"],
                "name": a["name"],
                "category": a.get("category") or "Armor",
                "armor": a.get("armor") or "0",
                "armorcapacity": a.get("armorcapacity") or "",
                "avail": a.get("avail") or "",
                "cost": a.get("cost") or "0",
                "minrating": int(a.get("minrating") or 0),
                "maxrating": int(a.get("maxrating") or 0),
                "additive": bool(a.get("additive")),
                "addmodcategories": list(a.get("addmodcategories") or []),
                "has_wireless": bool(a.get("wirelessbonus")),
                "source": a.get("source") or "",
                "page": a.get("page") or "",
            }
            for a in raw.get("armor") or []
        ],
        "armor_mods": [
            {
                "id": a["id"],
                "name": a["name"],
                "category": a.get("category") or "General",
                "armor": a.get("armor") or "0",
                "armorcapacity": a.get("armorcapacity") or "",
                "avail": a.get("avail") or "",
                "cost": a.get("cost") or "0",
                "minrating": int(a.get("minrating") or 0),
                "maxrating": int(a.get("maxrating") or 0),
                "purchasable": bool(a.get("purchasable")),
                "unique": a.get("unique") or "",
                "required_names": list(a.get("required_names") or []),
                "required_mods": list(a.get("required_mods") or []),
                "has_wireless": bool(a.get("wirelessbonus")),
                "source": a.get("source") or "",
                "page": a.get("page") or "",
            }
            for a in raw.get("armor_mods") or []
            if a.get("purchasable")
        ],
        "weapons": [
            {
                "id": w["id"],
                "name": w["name"],
                "category": w.get("category") or "",
                "type": w.get("type") or "",
                "weapon_type": w.get("weapon_type") or "",
                "accuracy": w.get("accuracy") or "",
                "reach": w.get("reach") or "",
                "damage": w.get("damage") or "",
                "ap": w.get("ap") or "",
                "mode": w.get("mode") or "",
                "ammo": w.get("ammo") or "",
                "conceal": w.get("conceal") or "",
                "range": w.get("range") or "",
                "alt_range": w.get("alt_range") or "",
                "mounts": list(w.get("mounts") or []),
                "avail": w.get("avail") or "",
                "cost": w.get("cost") or "0",
                "source": w.get("source") or "",
                "page": w.get("page") or "",
                "from_gear": bool(w.get("from_gear")),
                "add_gear_id": w.get("add_gear_id") or "",
            }
            for w in raw.get("weapons") or []
            if not w.get("hidden")
        ],
        "weapon_accessories": [
            {
                "id": a["id"],
                "name": a["name"],
                "mounts": list(a.get("mounts") or []),
                "avail": a.get("avail") or "",
                "cost": a.get("cost") or "0",
                "purchasable": bool(a.get("purchasable")),
                "accuracy": a.get("accuracy") or "",
                "rc": a.get("rc") or "",
                "minrating": int(a.get("minrating") or 0),
                "maxrating": int(a.get("maxrating") or 0),
                "required": a.get("required") or {},
                "forbidden": a.get("forbidden") or {},
                "specialmodification": bool(a.get("specialmodification")),
                "special_modification_cost": int(a.get("special_modification_cost") or 0),
                "source": a.get("source") or "",
                "page": a.get("page") or "",
            }
            for a in raw.get("weapon_accessories") or []
        ],
        "gear": [
            {
                "id": c["id"],
                "name": c["name"],
                "category": c.get("category") or "",
                "cost": c.get("cost") or "0",
                "avail": c.get("avail") or "",
                "minrating": int(c.get("minrating") or 0),
                "maxrating": int(c.get("maxrating") or 0),
                "capacity": c.get("capacity") or "",
                "plugin": bool(c.get("plugin")),
                "requireparent": bool(c.get("requireparent")),
                "addoncategories": list(c.get("addoncategories") or []),
                "required_names": list(c.get("required_names") or []),
                "required_categories": list(c.get("required_categories") or []),
                "ammo_weapon_types": list(c.get("ammo_weapon_types") or []),
                "costfor": int(c.get("costfor") or 0),
                "weapon_details": c.get("weapon_details") or "",
                "add_weapon": c.get("add_weapon") or "",
                "add_weapon_id": c.get("add_weapon_id") or "",
                "needs_extra": bool(c.get("needs_extra")),
                "extra_kind": c.get("extra_kind") or "",
                "extra_options": gear_extra_options(c, raw.get("skills")),
                "source": c.get("source") or "",
                "page": c.get("page") or "",
            }
            for c in raw.get("gear") or []
        ],
        "lifestyles": [
            {
                "id": ls["id"],
                "name": ls["name"],
                "cost": int(ls.get("cost") or 0),
                "dice": int(ls.get("dice") or 0),
                "lp": int(ls.get("lp") or 0),
                "multiplier": int(ls.get("multiplier") or 100),
                "increment": ls.get("increment") or "month",
                "freegrids": [
                    {"name": str(g.get("name") or ""), "select": str(g.get("select") or "")}
                    for g in (ls.get("freegrids") or [])
                ],
                "source": ls.get("source") or "",
                "page": ls.get("page") or "",
            }
            for ls in raw.get("lifestyles") or []
        ],
        "lifestyle_qualities": [
            {
                "id": q["id"],
                "name": q["name"],
                "category": q.get("category") or "",
                "lp": int(q.get("lp") or 0),
                "cost": int(q.get("cost") or 0),
                "multiplier": int(q.get("multiplier") or 0),
                "allowed": list(q.get("allowed") or []),
                "allow_multiple": bool(q.get("allow_multiple")),
                "needs_extra": bool(q.get("needs_extra")),
                "source": q.get("source") or "",
                "page": q.get("page") or "",
            }
            for q in raw.get("lifestyle_qualities") or []
        ],
        "drugs": [
            {
                "id": item["id"],
                "name": item["name"],
                "category": item.get("category") or "",
                "cost": item.get("cost") or "0",
                "avail": item.get("avail") or "",
                "addoncategories": list(item.get("addoncategories") or []),
                "speed": item.get("drug_speed") or "",
                "vectors": list(item.get("drug_vectors") or []),
                "duration": item.get("drug_duration") or "",
                "effect": drug_effect_summary(item.get("drug_bonus") or []),
                "source": item.get("source") or "",
                "page": item.get("page") or "",
            }
            for item in raw.get("drugs") or []
        ],
        "drug_grades": [
            {
                "id": item["id"],
                "name": item["name"],
                "cost": item.get("cost") or "0",
                "avail": item.get("avail") or "",
                "required_categories": list(item.get("required_categories") or []),
                "source": item.get("source") or "",
                "page": item.get("page") or "",
            }
            for item in raw.get("drug_grades") or []
        ],
    }

File: frontend/lib/types/rows/gear.ts
// Armor, weapons, matrix devices, gear and drugs: row shapes listed in `Derived` (../derived.ts).

import type { Notice } from "@/lib/engine-notices";

export interface SpecialArmor {
  fire?: number;
  cold?: number;
  electricity?: number;
  radiation?: number;
  sonic?: number;
  toxin_contact?: number;
  toxin_ingestion?: number;
  toxin_inhalation?: number;
  toxin_injection?: number;
  pathogen_contact?: number;
  pathogen_ingestion?: number;
  pathogen_inhalation?: number;
  pathogen_injection?: number;
  immunities?: {
    toxin_contact?: boolean;
    toxin_inhalation?: boolean;
    pathogen_contact?: boolean;
    pathogen_inhalation?: boolean;
  };
}

export interface LimitModifier {
  limit: "physical" | "mental" | "social" | string;
  value: number;
  condition?: string;
  condition_label?: Notice | null;
  source?: string;
}

export interface InstalledArmorMod {
  id: string;
  mod_id: string;
  name: string;
  category?: string;
  parent_id?: string | null;
  included?: boolean;
  rating: number;
  rating_max: number;
  nuyen: number;
  capacity_cost?: number;
  armor?: string;
  unique?: string;
  wireless?: boolean;
  has_wireless?: boolean;
  avail?: string;
  source?: string;
  special_armor?: SpecialArmor;
  limit_modifiers?: LimitModifier[];
  /** Custom Fit (Stack) `<selectarmor>`: the mod names an armor to stack with */
  select_armor?: boolean;
  stack_with?: string;
}

export interface InstalledArmor {
  /** `Variable(lo-hi)`: the range the player prices it within. */
  cost_range?: [number, number] | null;
  id: string;
  armor_id: string;
  name: string;
  category: string;
  armor: string;
  armor_value: number;
  additive: boolean;
  /** `+N` a Custom Fit (Stack) piece adds to the armor it stacks with */
  armoroverride?: string;
  rating: number;
  rating_max: number;
  equipped: boolean;
  wireless?: boolean;
  has_wireless?: boolean;
  nuyen: number;
  avail?: string;
  source?: string;
  contributes?: number;
  armorcapacity?: string;
  addmodcategories?: string[];
  mods?: InstalledArmorMod[];
  /** gear carried in it (a Holster, a Medkit) */
  gear?: InstalledGear[];
  capacity_used?: number;
  capacity_max?: number;
}

export interface InstalledWeapon {
  id: string;
  weapon_id: string;
  name: string;
  category: string;
  type: string;
  weapon_type?: string;
  accuracy: string;
  /** `Physical` / `Physical-1` before the engine put the limit in. */
  accuracy_formula?: string;
  /** `({STR}+5)P` before the engine put the body's STR in. */
  damage_formula?: string;
  reach: string;
  damage: string;
  ap: string;
  mode: string;
  ammo: string;
  rc?: string;
  rc_total?: number;
  conceal?: string;
  range?: string;
  alt_range?: string;
  mounts?: string[];
  qty: number;
  nuyen: number;
  accessories?: InstalledWeaponAccessory[];
  ammo_gear?: InstalledGear[];
  loaded_ammo_id?: string;
  from_gear?: boolean;
  source_gear_id?: string;
  from_ware?: boolean;
  source_ware_id?: string;
  /** Which ware tab owns the implant that grants it: "cyberware" or "bioware". */
  ware_kind?: string;
  /** A shield: armour that is also a weapon, deleted by dropping the armour. */
  from_armor?: boolean;
  source_armor_id?: string;
  /** Born with it (a `<naturalweapon>` grant) — nothing to buy, install or drop. */
  natural?: boolean;
  natural_source?: string;
  useskill?: string;
  limb_str?: number | null;
  limb_agi?: number | null;
  mounted_on?: string;
  mounted_label?: string;
  focus_dice?: number;
  category_dice?: number;
  avail?: string;
  source?: string;
}

export interface InstalledWeaponAccessory {
  id: string;
  accessory_id: string;
  name: string;
  parent_id?: string | null;
  included?: boolean;
  mount?: string;
  rating?: number;
  rating_max?: number;
  nuyen: number;
  accuracy?: string;
  rc?: string;
  avail?: string;
  source?: string;
  specialmodification?: boolean;
  special_modification_cost?: number;
}

export interface InstalledCommlink {
  id: string;
  gear_id: string;
  name: string;
  category?: string;
  rating: number;
  rating_max: number;
  qty?: number;
  device_rating: number;
  /** from a plugged-in Attack / Stealth Dongle (DT p.61); 0 on a bare commlink */
  attack?: number;
  sleaze?: number;
  dataprocessing: number;
  firewall: number;
  nuyen: number;
  apps?: InstalledProgram[];
  avail?: string;
  source?: string;
}

export interface InstalledMatrixDevice {
  id: string;
  gear_id: string;
  name: string;
  category?: string;
  rating: number;
  rating_max: number;
  device_rating: number;
  attack?: number;
  sleaze?: number;
  dataprocessing: number;
  firewall: number;
  programs?: number;
  program_used?: number;
  program_max?: number;
  nuyen: number;
  avail?: string;
  source?: string;
  array?: number[];
  array_order?: string[];
  can_reorder?: boolean;
}

export interface InstalledOptics {
  id: string;
  gear_id: string;
  name: string;
  category: string;
  rating: number;
  rating_max: number;
  parent_id?: string | null;
  included?: boolean;
  plugin?: boolean;
  nuyen: number;
  capacity_cost?: number;
  capacity_used?: number;
  capacity_max?: number;
  /** what it takes of the armor it is carried in */
  armor_capacity?: number;
  /** on a row carried in armor: the list it lives in */
  bucket?: "gear" | "optics" | "sensors";
  addoncategories?: string[];
  requireparent?: boolean;
  device_rating?: number;
  avail?: string;
  source?: string;
}

export interface InstalledGear extends InstalledOptics {
  /** `Variable(lo-hi)`: the range the player prices it within. */
  cost_range?: [number, number] | null;
  /** a Custom Item's own name */
  custom_name?: string;
  label?: string;
  qty: number;
  extra?: string;
  needs_extra?: boolean;
  extra_kind?: string;
  /** the translation kind of a pick — see `gear_extra_skill_kind` */
  extra_skill_kind?: "skill" | "knowledge_skill" | "";
  required_names?: string[];
  required_categories?: string[];
  ammo_weapon_types?: string[];
  costfor?: number;
  add_weapon?: string;
  add_weapon_id?: string;
  loaded?: boolean;
  is_drug?: boolean;
  active?: boolean;
  drug_speed?: string;
  drug_vectors?: string[];
  drug_duration?: string;
  drug_effect?: Notice[];
  /** `<addgear>`: the quality that handed this row over (Dead SIN). */
  granted_by?: string;
}

export interface InstalledProgram {
  /** `Variable(lo-hi)`: the range the player prices it within. */
  cost_range?: [number, number] | null;
  id: string;
  gear_id: string;
  name: string;
  category: string;
  rating: number;
  rating_max: number;
  parent_id?: string | null;
  /** came with its host (a Nixdorf Sekretar's Agent): free */
  included?: boolean;
  extra?: string;
  label?: string;
  needs_extra?: boolean;
  extra_kind?: string;
  /** the translation kind of a pick — see `gear_extra_skill_kind` */
  extra_skill_kind?: "skill" | "knowledge_skill" | "";
  nuyen: number;
  program_host?: string;
  avail?: string;
  source?: string;
}

export interface ActiveDrug {
  name: string;
  category: string;
  speed?: string;
  vectors?: string[];
  duration?: Notice | null;
  effect?: Notice[];
}

export interface CustomDrugComponentRow {
  component_id: string;
  name: string;
  category: string;
  level: number;
}

/** A mixed drug as the engine resolved it: the totals a bought drug would read
 *  off its catalog entry, summed from the components instead. */
export interface CustomDrug {
  id: string;
  name: string;
  grade: string;
  qty: number;
  active: boolean;
  nuyen: number;
  avail: string;
  addiction_rating: number;
  addiction_threshold: number;
  crash_damage: number;
  /** Onset in seconds; under 3 is immediate (a Combat Turn). */
  speed: number;
  /** Duration in seconds; 0 when no component gives one. */
  duration: number;
  infos: string[];
  components: CustomDrugComponentRow[];
  effect?: Notice[];
  source?: string;
  page?: string;
}
